File: ring_ass4/ring.py
class Process:
    def __init__ (self, process_id, is_active=True):
        self.id = process_id
        self.is_active = is_active
        self.next_process = None

    def __repr__ (self):
        return f"Process(ID: {self.id}, Active: {self.is_active})"
    
def elect_leader (processes, initiator_id):

    if not processes:
        return None
    
    initiator_index = -1
    for i, p in enumerate(processes):
        if p.id == initiator_id:
            initiator_index = i
            break

    if initiator_index == -1 or not processes(initiator_index).is_active:
        print(f"Initiator Process {initiator_id} not found or inactive")
        return None

    for i in range(len(processes)):
        processes[i].next_processes = processes[(i+1) % len(processes)]
    
    election_message = [processes[initiator_index].id]
    current_process = processes[initiator_index].next_process

    while current_process.id != initiator_id or len(election_message) < len([p for p in processes if p.is_active]):
        if current_process.is_active:
            print(f"Process {current_process.id} receives message: {election_message}")
            if current_process.id not in election_message:
                election_message.append(current_process.id)
            election_message.sort(reverse=True)
        else:
            print(f"Process {current_process.id} is inactive, message passed through.")

        current_process = current_process.next_process

        if current_process.id == initiator_id and len(election_message) == len([p for p in processes if p.is_active]):
            break

    leader_id = election_message[0]
    print(f"\nElection complete. Elected Leader: Process {leader_id}")
    return leader_id

File: ass4/ring/ring.java
import java.util.*;

class Process {
    int id;
    boolean isActive;
    Process next;

    public Process(int id, boolean isActive) {
        this.id = id;
        this.isActive = isActive;
        this.next = null;
    }

    @Override
    public String toString() {
        return "Process(ID: " + id + ", Active: " + isActive + ")";
    }
}

public class ring {

    public static Integer electLeader(List<Process> processes, int initiatorId) {
        if (processes == null || processes.isEmpty())
            return null;

        // Find initiator
        Process initiator = null;
        for (Process p : processes) {
            if (p.id == initiatorId) {
                initiator = p;
                break;
            }
        }

        if (initiator == null || !initiator.isActive) {
            System.out.println("Initiator Process " + initiatorId + " not found or inactive");
            return null;
        }

        // Connect processes in a ring
        int n = processes.size();
        for (int i = 0; i < n; i++) {
            processes.get(i).next = processes.get((i + 1) % n);
        }

        // Start election
        List<Integer> electionMessage = new ArrayList<>();
        electionMessage.add(initiator.id);

        Process current = initiator.next;

        System.out.println("Starting election from Process " + initiator.id);
        while (true) {
            if (current.isActive) {
                System.out.println("Process " + current.id + " receives message: " + electionMessage);
                if (!electionMessage.contains(current.id)) {
                    electionMessage.add(current.id);
                }
                // Sort descending to keep highest ID first
                electionMessage.sort(Collections.reverseOrder());
            } else {
                System.out.println("Process " + current.id + " is inactive, message passed through.");
            }

            current = current.next;

            // If message completes one full circle
            if (current.id == initiator.id) {
                break;
            }
        }

        int leaderId = electionMessage.get(0);
        System.out.println("\nElection complete. Elected Leader: Process " + leaderId);
        return leaderId;
    }

    public static void main(String[] args) {
        // Example processes
        List<Process> processes = new ArrayList<>();
        processes.add(new Process(1, true));
        processes.add(new Process(2, false));
        processes.add(new Process(3, true));
        processes.add(new Process(4, true));
        processes.add(new Process(5, true));

        electLeader(processes, 3);
    }
}
